events/quarkus/src/main/java/org/projectnessie/events/quarkus/delivery/StandardEventDelivery.java

```diff
@@ -0,0 +1,123 @@
+
+import io.vertx.core.Vertx;
+import java.time.Duration;
+import org.projectnessie.events.api.Event;
+import org.projectnessie.events.quarkus.config.QuarkusEventConfig;
+import org.projectnessie.events.service.EventService;
+import org.projectnessie.events.spi.EventSubscriber;
+
+/**
+ * A Non-blocking {@link RetriableEventDelivery} that executes delivery attempts directly on Vert.x
+ * event loop. Suitable only for subscribers that do not block the event loop.
+ */
+public class StandardEventDelivery extends RetriableEventDelivery {
+
+  private final Event event;
+  private final EventSubscriber subscriber;
+  private final QuarkusEventConfig.RetryConfig config;
+  private final Vertx vertx;
+  private RetriableEventDelivery self = this;
+
+  // For testing only: 0 = not finished, 1 = success, 2 = failure, 3 = rejected
+  static final byte STATE_NOT_FINISHED = 0;
+  static final byte STATE_SUCCESS = 1;
+  static final byte STATE_FAILURE = 2;
+  static final byte STATE_REJECTED = 3;
+  byte state = STATE_NOT_FINISHED;
+
+  public StandardEventDelivery(
+      Event event, EventSubscriber subscriber, QuarkusEventConfig.RetryConfig config, Vertx vertx) {
+    this.event = event;
+    this.subscriber = subscriber;
+    this.config = config;
+    this.vertx = vertx;
+  }
+
+  @Override
+  public void start() {
+    if (subscriber.accepts(event)) {
+      self.startAttempt(1, config.getInitialDelay(), null);
+    } else {
+      self.deliveryRejected();
+    }
+  }
+
+  @Override
+  protected void startAttempt(int currentAttempt, Duration nextDelay, Throwable previousError) {
+    try {
+      self.tryDeliver(currentAttempt);
+      self.deliverySuccessful(currentAttempt);
+    } catch (Exception e) {
+      self.attemptFailed(currentAttempt, nextDelay, addSuppressed(e, previousError));
+    }
+  }
+
+  @Override
+  protected void tryDeliver(int currentAttempt) {
+    EventService.notifySubscriber(subscriber, event);
+  }
+
+  @Override
+  protected void attemptFailed(int lastAttempt, Duration nextDelay, Throwable error) {
+    if (lastAttempt < config.getMaxAttempts()) {
+      self.scheduleRetry(lastAttempt, nextDelay, error);
+    } else {
+      self.deliveryFailed(lastAttempt, error);
+    }
+  }
+
+  @Override
+  protected void scheduleRetry(int lastAttempt, Duration nextDelay, Throwable lastError) {
+    vertx.setTimer(
+        nextDelay.toMillis(),
+        id -> self.startAttempt(lastAttempt + 1, config.getNextDelay(nextDelay), lastError));
+  }
+
+  @Override
+  protected void deliverySuccessful(int lastAttempt) {
+    state = STATE_SUCCESS;
+  }
+
+  @Override
+  protected void deliveryFailed(int lastAttempt, Throwable error) {
+    state = STATE_FAILURE;
+  }
+
+  @Override
+  protected void deliveryRejected() {
+    state = STATE_REJECTED;
+  }
+
+  @Override
+  protected final void setSelf(RetriableEventDelivery self) {
+    this.self = self;
+  }
+
+  @Override
+  protected final RetriableEventDelivery getSelf() {
+    return self;
```

Comment:
Looks like `self` is always `this` - is it needed?

Reply:
All the decorating impls do that, so that the outermost decorating impl "wins". For example:

```java
StandardEventDelivery standard = ...;
LoggingEventDelivery logging = new LoggingEventDelivery(standard, ev, subscr); // standard.self = logging
MetricsEventDelivery metrics = new MetricsEventDelivery(logging,  ev, subscr); // standard.self = metrics
TracingEventDelivery tracing = new TracingEventDelivery(metrics,  ev, subscr); // standard.self = tracing

tracing.start();
```

The call to `tracing.start()` would dispatch as follows:

```
tracing.start()
 -> metrics.start()
   -> logging.start()
     -> standard.start()
       -> standard.self.startAttempt()
         -> tracing.startAttempt()
           -> metrics.startAttempt()
             -> logging.startAttempt()
               -> standard.startAttempt()
                 -> standard.self.tryDeliver()
                   -> tracing.tryDeliver()
                     -> metrics.tryDeliver()
                       -> logging.tryDeliver()
                         -> standard.tryDeliver()
                           -> standard.self.deliverySuccessful()
 etc.
```

IOW it's a way to achieve C#-style delegates in Java :-)

Reply:
Yea - I hear it, somehow I fail to find the call to `setSelf` that does it. (I'm obviously blind - please bear with me)

Reply:
No worries, it's a lot to review.

The calls are here:

* https://github.com/projectnessie/nessie/blob/b7e8bda070a90df345115adf896f16011784c72f/events/quarkus/src/main/java/org/projectnessie/events/quarkus/delivery/LoggingEventDelivery.java#L52
* https://github.com/projectnessie/nessie/blob/b7e8bda070a90df345115adf896f16011784c72f/events/quarkus/src/main/java/org/projectnessie/events/quarkus/delivery/MetricsEventDelivery.java#L62
* https://github.com/projectnessie/nessie/blob/b7e8bda070a90df345115adf896f16011784c72f/events/quarkus/src/main/java/org/projectnessie/events/quarkus/delivery/TracingEventDelivery.java#L73

Reply:
Yea - seen those before.
The weird thing is, I added a breakpoint in `getSelf()` but couldn't see a call where `sefl != this` - now I see it being different.